src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+	 * connection list empty.
+	 */
+
+	entry = hash_search(ConnectionHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		entry->connections = NIL;
+	}
+
+	if (flags & CACHED_CONNECTION)
+	{
+		ListCell *connectionCell = NULL;
+
+		/* check connection cache for a connection that's not already in use */
+		foreach(connectionCell, entry->connections)
+		{
+			connection = (MultiConnection *) lfirst(connectionCell);
+
+			/* don't return claimed connections */
+			if (!connection->claimedExclusively)
+			{
+				if (flags & SESSION_LIFESPAN)
+				{
+					connection->sessionLifespan = true;
+				}
+				connection->activeInTransaction = true;
+
+				/*
+				 * Check whether we're right now allowed to open new
+				 * connections. A cached connection counts as new if it hasn't
+				 * been used in this transaction.
+				 *
+				 * FIXME: This should be removed soon, once all connections go
+				 * through this API.
+				 */
+				if (!connection->activeInTransaction &&
+					XactModificationLevel > XACT_MODIFICATION_DATA)
+				{
+					ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+									errmsg("cannot open new connections after the first "
+										   "modification command within a transaction")));
+				}
+
+				return connection;
+			}
+
+			/*
+			 * One could argue for erroring out when the connection is in a
+			 * failed state. But that'd be a bad idea for two reasons:
+			 *
+			 * 1) Generally starting a connection might fail, after calling
+			 *    this function, so calling code needs to handle that anyway.
+			 * 2) This might be used in code that transparently handles
+			 *    connection failure.
+			 */
+		}
+
+		/* no connection available, done if a new connection isn't desirable */
+		if (!(flags & NEW_CONNECTION))
+		{
+			return NULL;
+		}
+	}
+
+	/*
+	 * Check whether we're right now allowed to open new connections.
+	 *
+	 * FIXME: This should be removed soon, once all connections go through
+	 * this API.
+	 */
+	if (XactModificationLevel > XACT_MODIFICATION_DATA)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connections after the first modification "
+							   "command within a transaction")));
+	}
+
+	/*
+	 * Either no caching desired, or no pre-established, non-claimed,
+	 * connection present. Initiate connection establishment.
+	 */
+	connection = StartConnectionEstablishment(&key);
+
+	oldContext = MemoryContextSwitchTo(ConnectionContext);
+	entry->connections = lappend(entry->connections, connection);
+	MemoryContextSwitchTo(oldContext);
+
+	if (flags & SESSION_LIFESPAN)
+	{
+		connection->sessionLifespan = true;
+	}
+
+	connection->activeInTransaction = true;
+
+	return connection;
+}
+
+
+/*
+ * Synchronously finish connection establishment of an individual connection.
+ *
+ * TODO: Replace with variant waiting for multiple connections.
+ */
+void
+FinishConnectionEstablishment(MultiConnection *connection)
```

Comment:
I know we have some wonky calling order rules, but maybe break them here such that this can be just below StartConnectionEstablishment.